src/Nethermind/Nethermind.Blockchain/HistoryPruning/HistoryPruner.cs

```diff
@@ -0,0 +1,269 @@
+        IHistoryConfig historyConfig,
+        long secondsPerSlot,
+        ILogManager logManager)
+    {
+        _specProvider = specProvider;
+        _logger = logManager.GetClassLogger();
+        _blockTree = blockTree;
+        _receiptStorage = receiptStorage;
+        _blockStore = blockStore;
+        _chainLevelInfoRepository = chainLevelInfoRepository;
+        _metadataDb = metadataDb;
+        _historyConfig = historyConfig;
+        _enabled = historyConfig.Enabled;
+        _epochLength = secondsPerSlot * 32;
+        _minHistoryRetentionEpochs = specProvider.GenesisSpec.MinHistoryRetentionEpochs;
+
+        CheckConfig();
+        LoadDeletePointer();
+    }
+
+    public void OnBlockProcessorQueueEmpty(object? sender, EventArgs e)
+        => _ = TryPruneHistory(CancellationToken.None);
+
+    public async Task TryPruneHistory(CancellationToken cancellationToken)
+    {
+        if (!ShouldPruneHistory())
+        {
+            return;
+        }
+
+        lock (_pruneLock)
+        {
+            if (_pruneHistoryTask is not null && !_pruneHistoryTask.IsCompleted)
+            {
+                return;
+            }
+
+            _pruneHistoryTask = PruneHistory(cancellationToken);
+        }
+        await _pruneHistoryTask;
+    }
+
+    public long? CutoffBlockNumber
+    {
+        get
+        {
+            ulong cutoffTimestamp = CalculateCutoffTimestamp();
+            long? cutoffBlockNumber = null;
+            IEnumerable<Block> block = GetBlocksByNumber(_deletePointer, b =>
+            {
+                bool afterCutoff = b.Timestamp >= cutoffTimestamp;
+                if (afterCutoff)
+                {
+                    cutoffBlockNumber = b.Number;
+                }
+                return afterCutoff;
+            }, _ => { });
+            return cutoffBlockNumber;
+        }
+    }
+
+    private void CheckConfig()
+    {
+        if (_historyConfig.HistoryRetentionEpochs is not null &&
+            _historyConfig.HistoryRetentionEpochs < _minHistoryRetentionEpochs)
+        {
+            throw new HistoryPrunerException($"HistoryRetentionEpochs must be at least {_minHistoryRetentionEpochs}.");
+        }
+    }
+
+    private bool ShouldPruneHistory()
+    {
+        if (!_enabled)
+        {
+            return false;
+        }
+
+        ulong cutoffTimestamp = CalculateCutoffTimestamp();
+        return cutoffTimestamp > _lastPrunedTimestamp;
+    }
+
+    private async Task PruneHistory(CancellationToken cancellationToken)
+    {
+        if (_blockTree.Head is null)
+        {
+            return;
+        }
+
+        ulong cutoffTimestamp = CalculateCutoffTimestamp();
+
+        if (cutoffTimestamp <= _lastPrunedTimestamp)
+        {
+            return;
+        }
+
+        if (_logger.IsInfo) _logger.Info($"Pruning historical blocks up to timestamp {cutoffTimestamp}");
+
+        await Task.Run(() => PruneBlocksAndReceipts(cutoffTimestamp, cancellationToken), cancellationToken);
```

Comment:
Probably run it with `IBackgroundTaskScheduler` so that it does not affect attestation. This will be long for existing user.